src/linear_algebra/free_algebra.lean

```diff
@@ -0,0 +1,32 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Wieser
+-/
+import linear_algebra.basis
+import algebra.free_algebra
+
+import linear_algebra.finsupp_vector_space
+/-!
+# Linear algebra properties of `free_algebra R X`
+
+This file provides a `free_monoid X` basis on the `free_algebra R X`, and uses it to show the
+dimension of the algebra is the cardinality of `list X`
+-/
+namespace free_algebra
+
+lemma is_basis_free_monoid (R : Type*) (X : Type*) [comm_ring R] :
+  is_basis R (λ x : free_monoid X, free_monoid.lift (ι R) x) :=
+begin
+  convert (equiv_monoid_algebra_free_monoid.symm.to_linear_equiv : _ ≃ₗ[R] free_algebra R X)
+    .is_basis finsupp.is_basis_single_one,
+  ext x,
+  rw ←one_smul R (free_monoid.lift _ _),
+  exact (monoid_algebra.lift_single _ _ _).symm,
+end
+
+lemma dim_eq {K : Type*} {X : Type*} [field K] :
+  vector_space.dim K (free_algebra K X) = cardinal.mk (list X) :=
+(cardinal.lift_inj.mp (is_basis_free_monoid K X).mk_eq_dim).symm
```

Comment:
The cardinality of `list X` is: 1 if X is empty, aleph_0 if X is nonempty and finite, and the same as X if X is infinite (at least under AC). It's hard to imagine that `dim_eq` is useful in its current form without this result, but of course that doesn't mean we shouldn't have it. Do we know the dimensions of polynomial rings and multivariable polynomials rings over K?

Reply:
My memory was that mathlib already has a statement or two about the cardinality of lists, although I don't remember where they were

Reply:
There's https://github.com/leanprover-community/mathlib/blob/360fa07e88141f49f1741dc06e631e8c3133b405/src/field_theory/mv_polynomial.lean#L115 for the dimension of multivariable polynomials. I don't see a statement for polynomials.

In any case Kevin's suggestion can be added later as another PR.

Reply:
I'd prefer to just put this PR in as is - the main result is the basis, I just thought this lemma would be convenient, and will be a now contrast against the lower dimension of the exterior algebra when we eventually have that.